doc/man/briskcoin.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH BRISKCOIN "1" "October 2025" "briskcoin v30.0.0" "User Commands"
.SH NAME
briskcoin \- manual page for briskcoin v30.0.0
.SH SYNOPSIS
.B briskcoin
[\fI\,OPTIONS\/\fR] \fI\,COMMAND\/\fR...
.SH OPTIONS
.TP
\fB\-m\fR, \fB\-\-multiprocess\fR
Run multiprocess binaries briskcoin\-node, briskcoin\-gui.
.TP
\fB\-M\fR, \fB\-\-monolithic\fR
Run monolithic binaries briskcoind, briskcoin\-qt. (Default behavior)
.TP
\fB\-v\fR, \fB\-\-version\fR
Show version information
.TP
\fB\-h\fR, \fB\-\-help\fR
Show full help message
.SS "Commands:"
.TP
gui [ARGS]
Start GUI, equivalent to running 'briskcoin\-qt [ARGS]' or 'briskcoin\-gui [ARGS]'.
.TP
node [ARGS]
Start node, equivalent to running 'briskcoind [ARGS]' or 'briskcoin\-node [ARGS]'.
.TP
rpc [ARGS]
Call RPC method, equivalent to running 'briskcoin\-cli \fB\-named\fR [ARGS]'.
.TP
wallet [ARGS]
Call wallet command, equivalent to running 'briskcoin\-wallet [ARGS]'.
.TP
tx [ARGS]
Manipulate hex\-encoded transactions, equivalent to running 'briskcoin\-tx [ARGS]'.
.TP
help
Show full help message.
.SS "Additional less commonly used commands:"
.TP
bench [ARGS]
Run bench command, equivalent to running 'bench_briskcoin [ARGS]'.
.IP
chainstate [ARGS] Run briskcoin kernel chainstate util, equivalent to running 'briskcoin\-chainstate [ARGS]'.
test [ARGS]       Run unit tests, equivalent to running 'test_briskcoin [ARGS]'.
test\-gui [ARGS]   Run GUI unit tests, equivalent to running 'test_briskcoin\-qt [ARGS]'.
.SH COPYRIGHT
Copyright (C) 2009-2025 The Briskcoin Core developers

Please contribute if you find Briskcoin Core useful. Visit
<https://briskcoincore.org/> for further information about the software.
The source code is available from <https://github.com/briskcoin/briskcoin>.

This is experimental software.
Distributed under the MIT software license, see the accompanying file COPYING
or <https://opensource.org/license/MIT>
.SH "SEE ALSO"
briskcoin(1), briskcoind(1), briskcoin-cli(1), briskcoin-tx(1), briskcoin-wallet(1), briskcoin-util(1), briskcoin-qt(1)


doc/release-notes.md
3.0.0 Release Notes
====================

Briskcoin Core version 3.0.0 is now available from:

  <https://briskcoin.org>

This release includes new features, various bug fixes and performance
improvements, as well as updated translations.

Please report bugs using the issue tracker at GitHub:

  <https://github.com/briskcoin-project/briskcoin/issues>

Downgrade warning
------------------

Because release 1.0+ and later makes use of headers-first synchronization and
parallel block download (see further), the block files and databases are not
backwards-compatible with pre-1.0 versions of briskcoin Core or other software:

* Blocks will be stored on disk out of order (in the order they are
received, really), which makes it incompatible with some tools or
other programs. Reindexing using earlier versions will also not work
anymore as a result of this.

* The block index database will now hold headers for which no block is
stored on disk, which earlier versions won't support.

If you want to be able to downgrade smoothly, make a backup of your entire data
directory. Without this your node will need start syncing (or importing from
bootstrap.dat) anew afterwards. 

This does not affect wallet forward or backward compatibility.


How to Upgrade
==============

If you are running an older version, shut it down. Wait until it has completely
shut down (which might take a few minutes in some cases), then run the
installer (on Windows) or just copy over `/Applications/Briskcoin-Qt` (on macOS)
or `briskcoind`/`briskcoin-qt` (on Linux).

Upgrading directly from a version of Briskcoin Core that has reached its EOL is
possible, but it might take some time if the data directory needs to be migrated. Old
wallet versions of Briskcoin Core are generally supported.

Compatibility
==============

Briskcoin Core is supported and tested on operating systems using the
Linux Kernel 3.17+, macOS 13+, and Windows 10+. Briskcoin
Core should also work on most other Unix-like systems but is not as
frequently tested on them. It is not recommended to use Briskcoin Core on
unsupported systems.

Notable changes
===============

Policy
------

- The maximum number of potentially executed legacy signature operations in a
  single standard transaction is now limited to 2500. Signature operations in all
  previous output scripts, in all input scripts, as well as all P2SH redeem
  scripts (if there are any) are counted toward the limit. The new limit is
  assumed to not affect any known typically formed standard transactions. The
  change was done to prepare for a possible BIP54 deployment in the future. (#32521)

- `-datacarriersize` is increased to 100,000 by default, which effectively uncaps
  the limit (as the maximum transaction size limit will be hit first). It can be
  overridden with `-datacarriersize=83` to revert to the limit enforced in previous
  versions. (#32406)

- Multiple data carrier (OP_RETURN) outputs in a transaction are now permitted for
  relay and mining. The `-datacarriersize` limit applies to the aggregate size of
  the scriptPubKeys across all such outputs in a transaction, not including the
  scriptPubKey size itself. (#32406)

- The minimum block feerate (`-blockmintxfee`) has been changed to 0.001 satoshi per
  vB. It can still be changed using the configuration option. This option can be used
  by miners to set a minimum feerate on packages added to block templates. (#33106)

- The default minimum relay feerate (`-minrelaytxfee`) and incremental relay feerate
  (`-incrementalrelayfee`) have been changed to 0.1 satoshis per vB. They can still
  be changed using their respective configuration options, but it is recommended to
  change both together if you decide to do so. (#33106)

  Other minimum feerates (e.g. the dust feerate, the minimum returned by the fee
  estimator, and all feerates used by the wallet) remain unchanged. The mempool minimum
  feerate still changes in response to high volume.

  Note that unless these lower defaults are widely adopted across the network, transactions
  created with lower fee rates are not guaranteed to propagate or confirm. The wallet
  feerates remain unchanged; `-mintxfee` must be changed before attempting to create
  transactions with lower feerates using the wallet. (#33106)

P2P and network changes
-----------------------

- Opportunistic 1-parent-1-child package relay has been improved to handle
  situations when the child already has unconfirmed parent(s) in the mempool.
  This means that 1p1c packages can be accepted and propagate, even if they are
  connected to broader topologies: multi-parent-1-child (where only 1 parent
  requires fee-bumping), grandparent-parent-child (where only parent requires
  fee-bumping) etc. (#31385)

- The transaction orphanage, which holds transactions with missing inputs temporarily
  while the node attempts to fetch its parents, now has improved Denial of Service protections.
  Previously, it enforced a maximum number of unique transactions (default 100,
  configurable using `-maxorphantx`). Now, its limits are as follows: the number of
  entries (unique by wtxid and peer), plus each unique transaction's input count divided
  by 10, must not exceed 3,000. The total weight of unique transactions must not exceed
  `404,000` Wu multiplied by the number of peers. (#31829)

- The `-maxorphantx` option no longer has any effect, since the orphanage no longer
  limits the number of unique transactions. Users should remove this configuration
  option if they were using it, as the setting will cause an error in future versions
  when it is no longer recognized. (#31829)

New `briskcoin` command
---------------------

- A new `briskcoin` command line tool has been added to make features more discoverable
  and convenient to use. The `briskcoin` tool just calls other executables and does not
  implement any functionality on its own. Specifically `briskcoin node` is a synonym for

src/briskcoin.cpp
// Copyright (c) 2025 The Briskcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include <briskcoin-build-config.h> // IWYU pragma: keep

#include <clientversion.h>
#include <util/fs.h>
#include <util/exec.h>
#include <util/strencodings.h>
#include <util/translation.h>

#include <iostream>
#include <string>
#include <tinyformat.h>
#include <vector>

const TranslateFn G_TRANSLATION_FUN{nullptr};

static constexpr auto HELP_USAGE = R"(Usage: %s [OPTIONS] COMMAND...

Options:
  -m, --multiprocess     Run multiprocess binaries briskcoin-node, briskcoin-gui.
  -M, --monolithic       Run monolithic binaries briskcoind, briskcoin-qt. (Default behavior)
  -v, --version          Show version information
  -h, --help             Show full help message

Commands:
  gui [ARGS]     Start GUI, equivalent to running 'briskcoin-qt [ARGS]' or 'briskcoin-gui [ARGS]'.
  node [ARGS]    Start node, equivalent to running 'briskcoind [ARGS]' or 'briskcoin-node [ARGS]'.
  rpc [ARGS]     Call RPC method, equivalent to running 'briskcoin-cli -named [ARGS]'.
  wallet [ARGS]  Call wallet command, equivalent to running 'briskcoin-wallet [ARGS]'.
  tx [ARGS]      Manipulate hex-encoded transactions, equivalent to running 'briskcoin-tx [ARGS]'.
  help           Show full help message.
)";

static constexpr auto HELP_FULL = R"(
Additional less commonly used commands:
  bench [ARGS]      Run bench command, equivalent to running 'bench_briskcoin [ARGS]'.
  chainstate [ARGS] Run briskcoin kernel chainstate util, equivalent to running 'briskcoin-chainstate [ARGS]'.
  test [ARGS]       Run unit tests, equivalent to running 'test_briskcoin [ARGS]'.
  test-gui [ARGS]   Run GUI unit tests, equivalent to running 'test_briskcoin-qt [ARGS]'.
)";

static constexpr auto HELP_SHORT = R"(
Run '%s help' to see additional commands (e.g. for testing and debugging).
)";

struct CommandLine {
    bool use_multiprocess{false};
    bool show_version{false};
    bool show_help{false};
    std::string_view command;
    std::vector<const char*> args;
};

CommandLine ParseCommandLine(int argc, char* argv[]);
static void ExecCommand(const std::vector<const char*>& args, std::string_view argv0);

int main(int argc, char* argv[])
{
    try {
        CommandLine cmd{ParseCommandLine(argc, argv)};
        if (cmd.show_version) {
            tfm::format(std::cout, "%s version %s\n%s", CLIENT_NAME, FormatFullVersion(), FormatParagraph(LicenseInfo()));
            return EXIT_SUCCESS;
        }

        std::string exe_name{fs::PathToString(fs::PathFromString(argv[0]).filename())};
        std::vector<const char*> args;
        if (cmd.show_help || cmd.command.empty()) {
            tfm::format(std::cout, HELP_USAGE, exe_name);
            if (cmd.show_help) {
                tfm::format(std::cout, HELP_FULL);
                return EXIT_SUCCESS;
            } else {
                tfm::format(std::cout, HELP_SHORT, exe_name);
                return EXIT_FAILURE;
            }
        } else if (cmd.command == "gui") {
            args.emplace_back(cmd.use_multiprocess ? "briskcoin-gui" : "briskcoin-qt");
        } else if (cmd.command == "node") {
            args.emplace_back(cmd.use_multiprocess ? "briskcoin-node" : "briskcoind");
        } else if (cmd.command == "rpc") {
            args.emplace_back("briskcoin-cli");
            // Since "briskcoin rpc" is a new interface that doesn't need to be
            // backward compatible, enable -named by default so it is convenient
            // for callers to use a mix of named and unnamed parameters. Callers
            // can override this by specifying -nonamed, but should not need to
            // unless they are passing string values containing '=' characters
            // as unnamed parameters.
            args.emplace_back("-named");
        } else if (cmd.command == "wallet") {
            args.emplace_back("briskcoin-wallet");
        } else if (cmd.command == "tx") {
            args.emplace_back("briskcoin-tx");
        } else if (cmd.command == "bench") {
            args.emplace_back("bench_briskcoin");
        } else if (cmd.command == "chainstate") {
            args.emplace_back("briskcoin-chainstate");
        } else if (cmd.command == "test") {
            args.emplace_back("test_briskcoin");
        } else if (cmd.command == "test-gui") {
            args.emplace_back("test_briskcoin-qt");
        } else if (cmd.command == "util") {
            args.emplace_back("briskcoin-util");
        } else {
            throw std::runtime_error(strprintf("Unrecognized command: '%s'", cmd.command));
        }
        if (!args.empty()) {
            args.insert(args.end(), cmd.args.begin(), cmd.args.end());
            ExecCommand(args, argv[0]);
        }
    } catch (const std::exception& e) {
        tfm::format(std::cerr, "Error: %s\nTry '%s --help' for more information.\n", e.what(), argv[0]);
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

CommandLine ParseCommandLine(int argc, char* argv[])
{
    CommandLine cmd;
    cmd.args.reserve(argc);
    for (int i = 1; i < argc; ++i) {
        std::string_view arg = argv[i];
        if (!cmd.command.empty()) {
            cmd.args.emplace_back(argv[i]);
        } else if (arg == "-m" || arg == "--multiprocess") {
            cmd.use_multiprocess = true;
        } else if (arg == "-M" || arg == "--monolithic") {
            cmd.use_multiprocess = false;
        } else if (arg == "-v" || arg == "--version") {
            cmd.show_version = true;
        } else if (arg == "-h" || arg == "--help" || arg == "help") {
            cmd.show_help = true;
        } else if (arg.starts_with("-")) {
            throw std::runtime_error(strprintf("Unknown option: %s", arg));
        } else if (!arg.empty()) {
            cmd.command = arg;
        }
    }
    return cmd;
}

//! Execute the specified briskcoind, briskcoin-qt or other command line in `args`
//! using src, bin and libexec directory paths relative to this executable, where
//! the path to this executable is specified in `wrapper_argv0`.
//!
//! @param args Command line arguments to execute, where first argument should
//!             be a relative path to a briskcoind, briskcoin-qt or other executable
//!             that will be located on the PATH or relative to wrapper_argv0.
//!
//! @param wrapper_argv0 String containing first command line argument passed to
//!                      main() to run the current executable. This is used to
//!                      help determine the path to the current executable and
//!                      how to look for new executables.
//
//! @note This function doesn't currently print anything but can be debugged
//! from the command line using strace or dtrace like:
//!
//!     strace -e trace=execve -s 10000 build/bin/briskcoin ...
//!     dtrace -n 'proc:::exec-success  /pid == $target/ { trace(curpsinfo->pr_psargs); }' -c ...
static void ExecCommand(const std::vector<const char*>& args, std::string_view wrapper_argv0)
{
    // Construct argument string for execvp
    std::vector<const char*> exec_args{args};
    exec_args.emplace_back(nullptr);

    // Try to call ExecVp with given exe path.
    auto try_exec = [&](fs::path exe_path, bool allow_notfound = true) {
        std::string exe_path_str{fs::PathToString(exe_path)};
        exec_args[0] = exe_path_str.c_str();
        if (util::ExecVp(exec_args[0], (char*const*)exec_args.data()) == -1) {
            if (allow_notfound && errno == ENOENT) return false;
            throw std::system_error(errno, std::system_category(), strprintf("execvp failed to execute '%s'", exec_args[0]));
        }
        throw std::runtime_error("execvp returned unexpectedly");
    };

    // Get the wrapper executable path.
    const fs::path wrapper_path{util::GetExePath(wrapper_argv0)};

    // Try to resolve any symlinks and figure out the directory containing the wrapper executable.
    std::error_code ec;
    fs::path wrapper_dir{fs::weakly_canonical(wrapper_path, ec)};
    if (wrapper_dir.empty()) wrapper_dir = wrapper_path; // Restore previous path if weakly_canonical failed.
    wrapper_dir = wrapper_dir.parent_path();

    // Get path of the executable to be invoked.
    const fs::path arg0{fs::PathFromString(args[0])};

    // Decide whether to fall back to the operating system to search for the
    // specified executable. Avoid doing this if it looks like the wrapper
    // executable was invoked by path, rather than by search, to avoid
    // unintentionally launching system executables in a local build.
    // (https://github.com/briskcoin/briskcoin/pull/31375#discussion_r1861814807)
    const bool fallback_os_search{!fs::PathFromString(std::string{wrapper_argv0}).has_parent_path()};

    // If wrapper is installed in a bin/ directory, look for target executable
    // in libexec/
    (wrapper_dir.filename() == "bin" && try_exec(fs::path{wrapper_dir.parent_path()} / "libexec" / arg0.filename())) ||
#ifdef WIN32
    // Otherwise check the "daemon" subdirectory in a windows install.
    (!wrapper_dir.empty() && try_exec(wrapper_dir / "daemon" / arg0.filename())) ||
#endif
    // Otherwise look for target executable next to current wrapper
    (!wrapper_dir.empty() && try_exec(wrapper_dir / arg0.filename(), fallback_os_search)) ||
    // Otherwise just look on the system path.
    (fallback_os_search && try_exec(arg0.filename(), false));
}
